Feat/nomad-jobspec canary promote releaser

paladin-devops opened the pull request:
This PR introduces a releaser for Nomad, which promotes a canary deployment. Supporting canaries in releases for Nomad bring canary/blue-green deployment strategies into Waypoint's lifecycle (#2817).

This is still very much a work in progress. Although I've successfully promoted a Nomad canary deployment with this plugin, I still intend to fully implement the release manager for the releaser, the destroyer (though this will likely be redundant with the Nomad platform destroyer), and there are other configuration options/changes I would like to include:

- [x] Specify particular Nomad task groups to promote (by default this promotes all)
- [x] A Consul URL, or IP/port combo if Consul isn't used
- [x] Add canary option to the Nomad platform plugin
- [x] Support generation ID changes in releaser with nomad-jobspec platform deployment
- [x] Evaluate if scaling should be supported
- [x] Evaluate if failing a canary deployment should be supported
- [x] Evaluate if reverting to a previous deployment should be supported

This is my first attempt at a plugin beyond the tutorial, and I look forward to any feedback while I continue with this draft!

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/waypoint?pullRequest=2938)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/waypoint?pullRequest=2938)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)


---
**1** out of **2** committers have signed the CLA.

- [x] paladin-devops
- [ ] josephrajewski


---
**josephrajewski** seems not to be a GitHub user.
You need a GitHub account to be able to sign the CLA. If you already have a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/waypoint?pullRequest=2938).</sub>

paladin-devops commented:
Since I started this draft, I worked through implementing the resource manager for the Nomad platform releaser. I also added the ability to supply configurations to the Nomad platform deployer for a Nomad job's `update` stanza (now commented out). I removed that change because only after doing that did it occur to me that updating an existing deployment contrasts with Waypoint's opinionated workflow for Nomad, which, in part, is to create a new job upon deployment - not update an existing one. Therefore, I may very well revert all the changes made to the Nomad platform releaser, unless the opinionated workflow changes for Nomad in the direction of supporting a canary workflow using Nomad's built-in features.

Since then, I pivoted to focus on implementing a separate releaser, still for Nomad canaries, but specifically for the `nomad-jobspec` platform deployer, where much of the work I did for the platform releaser was portable. The implementation differs here in that `jobspec` is a plugin for mutable deployments and implements the [Generation](https://github.com/hashicorp/waypoint-plugin-sdk/blob/main/component/component.go#L334) interface. While doing this I also implemented the Resource Manager for `nomad-jobspec`:
![image](https://user-images.githubusercontent.com/83741749/151759241-3c181fca-d86e-4e2a-a081-36c1d27853af.png)

In order to permit canaries to work, the Generation implementation had to be tweaked. If a Nomad jobspec being deployed is a canary deployment, it [gets a UUID for its generation ID](
https://github.com/hashicorp/waypoint/pull/1298#issue-853824618). If it is not, the job's generation ID will be the Nomad job's ID, which is what previously it was always set to. In order to allow a release to occur with a mutable platform deployer, the generation ID cannot be the same as the last deployment. I have been wondering if to support canaries, that perhaps a canary flag within the Generation type could be useful, but that would require some deeper core changes. I'm also looking into implementing the Generation interface for this releaser.

Open issues:
- [ ] `waypoint deploy` has the flag `-release` set to true by default; however this almost always will kick off a release before the canaries become healthy, which is a pre-requisite to promotion for Nomad - this must be somehow overridden (if possible) or a delay before the initiation of the release should be considered, possibly by checking the healthy deadline of the update
- [ ] `waypoint release` prunes deployments by default. Even though the problem of allowing a release is taken care of by the Generation changes mentioned, the release operation itself will actually by default delete the job from which it just promoted the canaries as it prunes all but 1 deployment by default. So by the end of the release operation, the status retrieval which occurs leads to an EOF, as the job no longer exists from which to get a status.

Regarding the other open topics from my first comment, I still intend to determine a URL based on a Consul service in the job (specifics TBD), and add support for failing or reverting the canary deployment to the releaser.

The happy path at this point is to have a jobspec with canaries configured, run `waypoint deploy -release=false` (run it a second time if it's your first time deploying the job - the "update" stanza in your job does nothing if it is the very first deployment of the job), and then run `waypoint release -prune=false`.

CLI:
![image](https://user-images.githubusercontent.com/83741749/151762357-dadd6931-56b5-49b8-a310-b8171a409d9a.png)

UI:
![image](https://user-images.githubusercontent.com/83741749/151762565-7cd30851-ce29-4b0f-9a60-36de9eca1da3.png)

paladin-devops commented:
At this point, I'm going to leave out scaling and revert functionality from this PR. Promoting or failing a Nomad deployment are the operations most relevant to canaries. I considered adding revert support moreso than scaling, but to fully support the job revert Nomad API requires supplying a Consul ([though not yet implemented on the Nomad API side](https://github.com/hashicorp/nomad/blob/main/api/jobs.go#L435)) and Vault token, which would add greater complexity to the release operation being performed by Waypoint.

Failing a canary deployment is supported with the `fail_deployment` option, and promoting a specific group is supported with the `groups` config option.

Example usage:
```hcl
  release {
    use "nomad-jobspec-canary" {
      groups = [
        "test" //only promotes the job taskgroup "test", any others will be ignored (even if they have canaries)
      ]
      fail_deployment = false //if true, fails the active canary deployment
    }
  }
```

To hopefully wrap this up I'm going to now focus on the release URL, deployment pruning and canary-promotion-before-healthy issues mentioned in my previous two comments. I foresee the Consul URL as the release URL presenting more issues because if the app is using dynamic ports, then the URL with the Consul service name and a port appended to the end wouldn't be possible. In the meantime I may lean on the idea of selecting from the deployed job a random alloc's IP/port combo until the path to integration with Consul here (or with a load balancer that integrates with Consul like Traefik or Fabio) becomes clear.

briancain commented:
Thanks for all the details and keeping us in the loop @paladin-devops - Let us know when you're ready for a team review! For now, we'll give you the space to continue to work on this PR while it's still in draft form :) but wanted to say feel free to ping us when you are feeling like this is ready.

paladin-devops commented:
Thanks a lot @briancain, I appreciate the offer! I'm moving this PR out of the draft phase now.

Regarding the open points since my last comment, I have made the decision to configure the URL based on meta supplied by the user. I felt that this was acceptable (not ideal) since we ask the user to provide the PORT env var in the jobspec to opt-in to the URL service for the platform plugin. Supplying the metadata is also optional, if it is not set, the URL simply won't be displayed or logged by the releaser. I was making progress on the implementation of selecting a random Nomad host IP and port, but without requiring the user to specify which port label to use from the job to source the port from, picking the right port label/host port for the application would be a gamble. Adding this additional meta field will at least add certainty to the validity of the URL for the user. If this releaser were creating new resources which in-turn created a URL, like an ingress in k8s, it would be harder to argue that Waypoint doesn't need to supply that URL.

```hcl
  // This format is HCL1, but it conforms to the format of the meta for the deployment ID being set during the Deploy phase, waypoint.hashicorp.com/id 
  meta = {
    "waypoint.hashicorp.com/release_url" = "abc.com"
  }
```

```
waypoint release -prune=false

» Releasing nomad-promoter-test...
  Performing operation locally
✓ Allocation "875b2ace-7d8c-4242-183c-66b296ea7746" created: node "444786b0-8beb-ab13-7d95-4a8b3d04028a", group "test"
✓ Evaluation status changed: "pending" -> "complete"
✓ Evaluation "cec6c200-cf81-2cb8-c0e9-1a0536a4bd23" finished with status "complete"
✓ Release successfully rolled out!

  Performing operation locally
✓ Finished building report for Nomad platform
✓ Finished building report for Nomad job resource

Release URL: abc.com
```
UI:
![image](https://user-images.githubusercontent.com/83741749/152667046-ed50c334-09d8-4df5-960c-4f64eecdb87b.png)

Regarding the Consul service approach, using the Consul service's FQDN wouldn't be sufficient without a port, which again could be dynamic, and would vary from host to host, assuming multiple allocs and multiple Nomad clients on which the job's allocs could land. As a later enhancement, it may make sense to use a tag to uniquely identify one instance of the service, and then build a host name off of that for Waypoint to use, such as `waypoint_tag.app.service.dc1.consul`, and derive the port # from Nomad using that single instance. However that would require additional tweaks within the user-provided jobspec, probable integration with the Consul API, and would be the "release URL" for just one instance. I had other wily ideas on the topic of getting a URL, but for the scope of implementing a canary releaser, the meta approach will do for now.

Regarding the canary promotion before the allocs are healthy issue, I added a loop to cycle through the task groups to be promoted until the canaries are healthy.

Finally, with the pruning issue, after much deliberation and experimentation I added a note to the documentation for this plugin to recommend that the user supply the `-prune=false` flag. I tried many different ways to prevent the Nomad job for the just-released app from being pruned with the default core workflow of pruning old deployments, but to no avail. I decided to take this approach and not attempt to change the core workflow after reading the [note in the releaser lifecycle docs about pruning and the future plans for it](https://www.waypointproject.io/docs/lifecycle/release#deployment-pruning).

I fun working on this! Looking forward to getting it polished and merged up. :)

krantzinator commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,547 @@
+	groupsToPromote := make([]string, len(deploy.TaskGroups))
+	for taskGroupName, taskGroup := range deploy.TaskGroups {
+		if r.config.Groups != nil {
+			if isElementExist(r.config.Groups, taskGroupName) && taskGroup.DesiredCanaries != 0 {
```
You use `> 0` when updating the task group's canary, so for consistency, can we check `>` here too? Or is there a specific reason for checking `!= 0`?
```suggestion
			if isElementExist(r.config.Groups, taskGroupName) && taskGroup.DesiredCanaries != 0 {
```

paladin-devops replied:
@krantzinator `> 0` makes more sense here, thanks!

krantzinator commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,547 @@
+				groupsToPromote = append(groupsToPromote, taskGroupName)
+				continue
+			}
+		} else if taskGroup.DesiredCanaries != 0 {
+			canaryDeployment = true
+			// if no groups to promote are specified in the config, promote all groups
+			// that have canaries
+			groupsToPromote = append(groupsToPromote, taskGroupName)
+		}
```
Is this check supposed to be here, or just L123-L127 (I'm not sure what it's checking that wasn't already checked above)? Does this have the potential of a `nil` reference since we aren't checking for the existence of the task group here?

paladin-devops replied:
@krantzinator this check is here in case the user did not provide any task groups in the releaser's config.

Before implementing the ability to specify particular task groups to be promoted in the releaser config, I was using the `PromoteAll` Nomad API method. So when hitting this else block, to avoid adding a task group which has no canary allocs (`DesiredCanaries == 0`) to the `groupsToPromote` slice, we verify that it has canaries. If not, we continue to the next iteration, and if none of the task groups have canaries, we exit early (L135-L138).

So in effect, the logic of this for loop & hitting this else block is "if there are no groups specified in the releaser, promote every group that has canaries".

krantzinator reviewed (commented):
Just a few small comments and questions for now! Let me know if you need help on this docs generation as well. Looks great!

tgross commented on builtin/nomad/jobspec/platform.go:
```diff
@@ -46,86 +53,246 @@ func (p *Platform) GenerationFunc() interface{} {
+	})
+	jobResource.StateJson = string(stateJson)
+
+	if *job.Status == "running" {
+		jobResource.Health = sdk.StatusReport_READY
+		jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting ready!", state.Name)
+	} else if *job.Status == "queued" || *job.Status == "started" {
+		jobResource.Health = sdk.StatusReport_ALIVE
+		jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting alive!", state.Name)
+	} else if *job.Status == "completed" {
+		jobResource.Health = sdk.StatusReport_PARTIAL
+		jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting partially available!", state.Name)
+	} else if *job.Status == "failed" || *job.Status == "lost" {
+		jobResource.Health = sdk.StatusReport_DOWN
+		jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting down!", state.Name)
+	} else {
+		jobResource.Health = sdk.StatusReport_UNKNOWN
+		jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting unknown!", state.Name)
+	}
```
I think we might be co-mingling several different status values here. The status of a job is only one of [running, dead, or pending](https://github.com/hashicorp/nomad/blob/v1.2.5/nomad/structs/structs.go#L4021-L4025) (see also the [API docs](https://www.nomadproject.io/api-docs/jobs#status)). 

Whereas:
* An allocation's client status is one of [pending, running, complete, failed, or lost](https://github.com/hashicorp/nomad/blob/v1.2.5/nomad/structs/structs.go#L9322-L9328). 
* A deployment can be [running, paused, failed, successful, cancelled, pending, blocked, or unblocking](https://github.com/hashicorp/nomad/blob/v1.2.5/nomad/structs/structs.go#L8911-L8919)
* An eval can be [blocked, pending, complete, failed, canceled](https://github.com/hashicorp/nomad/blob/v1.2.5/nomad/structs/structs.go#L10370-L10376) (yes, those two objects spell canceled differently 😊 ). 
* I wasn't able to figure out which object has `queued` or `started`

The reason this granularity matters to a Waypoint user is that a Nomad job can be `running` but have a bunch of its allocations in various pending or failing states! So I suspect we want to aggregate information about the job, deployment, and its allocations and then present that as a top-level Waypoint job status.

paladin-devops replied:
@tgross thank you for advising on this! It looks like "queued" and "started" may have initially been sourced from the [TaskGroupSummary struct](https://github.com/hashicorp/nomad/blob/35c22bcb6c31d3605d9fe9f2a02cefee04e6c4b1/api/jobs.go#L1003-L1010), but I agree that it would make sense to do some kind of top-level status. Here's what I'm thinking...

If a deployment's status is "completed", skip the eval check and do alloc health check.

If a deployment's status is running, check eval ID returned. Since the eval may return a "blocked", "failed" or "canceled" status, that would allow us to skip the alloc evaluation and jump straight to a final unhealthy status.

If a deployment's status is "paused", "failed", or "blocked", early exit (skip alloc health check) with a final status.

If a deployment's status is "pending" or "unblocking", I'm a bit unsure how to proceed here... I'm thinking either provide an UNKNOWN status, or wait for some period of time and check again for a different status until we hit a timeout on the status check. 

For the alloc health check, if all allocs are "running", we go with "READY". If some are "running" and some are "pending" or "failed", we report "PARTIAL". If all are "lost" or "failed", we report "DOWN". I feel like if we have allocs in the "complete" state, then technically the job is "DOWN", but that could possibly be perceived as unhealthy to Waypoint users, even if the job ran to completion as intended (like a batch job).

Final status options for Waypoint, for reference:
- UNKNOWN: We cannot determine the status of the resource
- ALIVE: The resource exists but might not be ready to handle requests
- READY: The resource exists and is ready to handle requests
- DOWN: The resource is down and not responding
- MISSING: We're expecting it to exist, but it does not
- PARTIAL: Some resources in deployment are OK, others are not OK

tgross replied:
> If a deployment's status is "completed", skip the eval check and do alloc health check.

A deployment could be "complete" with failed-but-replaced allocations. Or you could complete the deploy and then immediately have an unrelated alloc failure. I'm not sure of the semantics the Waypoint user would want there in terms of the output of this command?

> If a deployment's status is running, check eval ID returned. Since the eval may return a "blocked", "failed" or "canceled" status, that would allow us to skip the alloc evaluation and jump straight to a final unhealthy status.

If the deployment is "running", the eval that led to that deployment can only be "complete". The scheduler may have _also_ spawned a new blocked eval if the plan was only partially applied. Ex. if you submit a job that needs 20 allocs but can only fit 18, we start the deployment for those 18 and then create a new "blocked" eval to re-plan the remaining 2 allocations once more resources become available. In that case it sounds like that'd be `PARTIAL`?

> If a deployment's status is "paused", "failed", or "blocked", early exit (skip alloc health check) with a final status.
>
> If a deployment's status is "pending" or "unblocking", I'm a bit unsure how to proceed here... I'm thinking either provide an UNKNOWN status, or wait for some period of time and check again for a different status until we hit a timeout on the status check.

A failed deployment is definitely a failure (`DOWN`?). 

A paused deployment has been manually paused by the operator with [`nomad deployment pause`](https://www.nomadproject.io/docs/commands/deployment/pause). Not sure what's the best Waypoint state to reflect that.

The "blocked", "pending", "unblocking" states are kind of exotic. Those are for [multi-region deployments](https://learn.hashicorp.com/tutorials/nomad/multiregion-deployments). I'm not sure if Waypoint supports federated Nomad regions at all yet, but if you got those you'd want to poll on them until they become terminal.

> For the alloc health check, if all allocs are "running", we go with "READY". If some are "running" and some are "pending" or "failed", we report "PARTIAL". If all are "lost" or "failed", we report "DOWN". I feel like if we have allocs in the "complete" state, then technically the job is "DOWN", but that could possibly be perceived as unhealthy to Waypoint users, even if the job ran to completion as intended (like a batch job).

That all seems reasonable to me.

Speaking of non-service jobs, you might also want to special case system jobs because they don't use deployments.

paladin-devops replied:
@tgross after further evaluation, I took a different direction here. It occurred to me that since Waypoint continuously polls for app status via the `resourceJobStatus` function (see: #1612), it may not be best for us to rely on the deployment or evaluation from the deployment event, since the information they provide will become stale after the deployment event. Because of that, in the update I made, we now check for the 3 job statuses, running/pending/dead, and only if the job is running do we dig deeper into the status of the job's allocs:

```go
		if running == currentJobVersionAllocs {
			jobResource.Health = sdk.StatusReport_READY
			jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting ready!", state.Name)
		} else if running > 0 && (complete > 0 || failed > 0 || pending > 0 || lost > 0) {
			jobResource.Health = sdk.StatusReport_PARTIAL
			jobResource.HealthMessage = fmt.Sprintf("Some allocations are running for Job %q!", state.Name)
		} else if (complete + pending + failed + lost) == currentJobVersionAllocs {
			jobResource.Health = sdk.StatusReport_DOWN
			jobResource.HealthMessage = fmt.Sprintf("No allocations for Job %q are running!", state.Name)
		}
```
https://github.com/hashicorp/waypoint/pull/2938/commits/608e787bb24121939dc603c72261ba7a6c0d15d8#diff-15311425528bec1ed88b93dfce5b4fa4ef9e23779e2d1f2229a55e976ccbb913R203-R230

paladin-devops replied:
@tgross I've updated the check the list of evals for an alloc, and if there's any blocked evals, but all allocs are running then the status becomes PARTIAL.

Eval check:
```go
	if *job.Status == "running" {
		// Get list of evaluations for job
		evals, _, err := jobClient.Evaluations(*job.ID, q)
		evalStatus := "unblocked"
		for _, eval := range evals {
			switch eval.Status {
			case structs.EvalStatusBlocked:
				evalStatus = "blocked"
				break
			case structs.EvalStatusPending:
				break
			case structs.EvalStatusComplete:
				break
			case structs.EvalStatusFailed:
				break
			case structs.EvalStatusCancelled:
				break
			default:
				evalStatus = "unknown"
			}
		}
```

Alloc check & eval check:
```go
		if running == currentJobVersionAllocs && evalStatus == "unblocked" {
			jobResource.Health = sdk.StatusReport_READY
			jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting ready!", state.Name)
		} else if running == currentJobVersionAllocs && evalStatus == "blocked" {
			jobResource.Health = sdk.StatusReport_PARTIAL
			jobResource.HealthMessage = fmt.Sprintf("Allocs for job %q are running, but there is at least one blocked evaluation!", state.Name)
		}
```

tgross commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,547 @@
+	jobs, _, err := jobClient.PrefixList(target.Name)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad jobs: %s", err.Error())
+	} else if target.Name != jobs[0].ID {
+		return status.Errorf(codes.Aborted, "Job not found: %s", err.Error())
+	}
```
In the Nomad CLI if the prefix isn't an exact match, we present a list of options to the user in the error message so they can refine the prefix match. Also, because `PrefixList` is a list API, Nomad doesn't return an error on empty results, so I believe this would panic with an index error if there's no results?

paladin-devops replied:
@tgross I added a check to make sure the list of jobs that the `PrefixList` actually has an element before checking the exactness of the job ID. Thanks!

& regarding presenting a list of options to the user, since the job ID we're checking against is actually the job ID from the Waypoint deploy phase, I think we should still fail here, but maybe output the name of the first one or two jobs returned in the list, saying something like "maybe you meant to promote the canary deployment for X job". Is this what you had in mind?

tgross replied:
Yeah, that's the sort of thing I had in mind.

tgross commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,547 @@
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad job: %s", err.Error())
+	}
+
+	st.Update("Getting latest deployments for job")
+	deploy, _, err := jobClient.LatestDeployment(*job.ID, q)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch latest deployment for Nomad job: %s", err.Error())
+	} else if deploy == nil {
+		st.Update("No active deployment for Nomad job")
+		return err
+	}
```
Because we have the deployment at this point, we might want to consider checking the initial deployment status before we go rummaging through the task groups. The deployment might have already failed, completed, been marked successful, etc.

Also, "latest deployment" means literally that and not necessarily the one you're expecting (this is admittedly a bit of a gotcha in the Nomad API!). When you submit a job:

* (1) The job is forwarded to the leader, and stored in the Nomad state store. At the same time, an evaluation is created.
* (2) A scheduler worker (not necessarily on the leader) dequeues the evaluation and creates a plan with the deployment. The plan is sent to the leader to be "applied".

So imagine if we submit a job via the API and then immediately hit `LatestDeployment` before the new deployment is created. We could end up with the deployment for the _previous_ version of the job, which would be confusing because we'd try to promote a possibly already-promoted deployment. So we'll want to match the version for this deployment against the job version we've just submitted.

paladin-devops replied:
@tgross I caught up to this comment after my response to your remarks on the status check, and I agree a similar "early exit" strategy could be applied here if the deployment fails/is unhealthy.

And thank you for the tip on the gotcha! I'll add in a check to verify that the job version returned by the job `Info` call matches up with the job version returned by the `LatestDeployment`.

tgross commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,547 @@
+			groupHealthy = false
+			for !groupHealthy {
+				if currentTaskGroupState.HealthyAllocs < len(currentTaskGroupState.PlacedCanaries) {
+					time.Sleep(5 * time.Second)
```
The top-level function this is in `resourceJobCreate` has a `context.Context`. Should we be using that here rather than having a fixed sleep without a way of breaking it? Is there any kind of backoff we should be using?

paladin-devops replied:
@tgross took some time, but I think I've got this covered! 

https://github.com/hashicorp/waypoint/pull/2938/files#diff-8465d427b10de27ab3337b90735cb0728c5f62a8ce0612476219fbb6aacc2574R148-R176

I have currently arbitrarily set the context deadline to 5 minutes here ([since it's the default](https://www.nomadproject.io/docs/job-specification/update#healthy_deadline)). Either in this PR or as an enhancement later I want to update this to pull the `HealthyDeadline` from the task group currently being assessed, and possibly also use child Go routines to concurrently check the task groups. Then if one of the task groups doesn't meet its healthy deadline before others don't, we may not have to wait extra time.

tgross reviewed (changes requested):
Hi @paladin-devops! Thanks for opening this PR! I'm a developer on the Nomad team so I'm going to admit I'm not as up-to-speed on Waypoint's code base. But I did take a pass through this looking at the Nomad-specific logic and I've left some comments for you.

paladin-devops commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,547 @@
+    }
+  }
+
+	group "app-gateway" {
```
I'm not sure what's going on with my Go formatting 😅 in my IDE (Atom) each of these appear to be aligned. I opened the file in Vim and the spacing being off was apparent so I just fixed it up there. Guess I'll do my dev work in Vim now 🤷🏻‍♂️

paladin-devops commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,576 @@
+				currentTaskGroupState = deploy.TaskGroups[group]
+				if currentTaskGroupState.HealthyAllocs < len(currentTaskGroupState.PlacedCanaries) {
+					select {
+					case <-time.After(5 * time.Second): // forces sleep for 5 seconds
```
@tgross thank you for the suggestion! I've updated the 5 second delay to instead use ticker to avoid the leak.

Also, I did not include another check for the deployment by calling the API at the top of the loop because earlier in the function [we get the deployment](https://github.com/hashicorp/waypoint/pull/2938/files/c80b5aca1fd787fa6f80e4a02715a82576e75c5a#diff-8465d427b10de27ab3337b90735cb0728c5f62a8ce0612476219fbb6aacc2574R111) for the first time. So unless the deployment status changed between that initial check and when we hit this loop, we still have a pretty current status of the deployment. I mostly did it this way to avoid a potentially likely unnecessary second call to the Nomad API, since we have a current deployment status already. However if you think making that second call is worth it to avoid hitting the select, I can definitely update it to get the deployment status again prior to hitting the select.

tgross commented on builtin/nomad/jobspec/platform.go:
```diff
@@ -2,23 +2,31 @@ package jobspec
 
 	"github.com/hashicorp/go-hclog"
 	"github.com/hashicorp/nomad/api"
+	"github.com/hashicorp/nomad/nomad/structs"
```
This package isn't a stable API. We import it into the Nomad Terraform provider and some of the plugins, and it's hugely painful and something we'd love to undo. Let's avoid pulling it into Waypoint and inflicting that pain on the Waypoint developers too 😀 

(That being said, although the alloc status consts are available in [`api/allocations.go`](https://github.com/hashicorp/nomad/blob/main/api/allocations.go#L17-L29) the eval and job status consts are not and absolutely should be. We'll need to fix this in the Nomad repo.)

paladin-devops replied:
Updated!

tgross commented on go.mod:
```diff
@@ -98,14 +99,14 @@ require (
-	google.golang.org/genproto v0.0.0-20210602131652-f16073e35f0c
-	google.golang.org/grpc v1.39.1
-	google.golang.org/protobuf v1.26.0
+	go.uber.org/goleak v1.1.12
```
This is kicking up some linter issues with the `go.sum`. I think the only reason `go.mod` is being updated here at all is because we pulled in `github.com/hashicorp/nomad/nomad/structs`? If we don't do that then we don't need to update this either.

tgross commented on builtin/nomad/jobspec/platform.go:
```diff
@@ -46,86 +54,314 @@ func (p *Platform) GenerationFunc() interface{} {
+	if *job.Status == structs.JobStatusRunning {
+		// Get list of evaluations for job
+		evals, _, err := jobClient.Evaluations(*job.ID, q)
+		evalStatus := "unblocked"
+		for _, eval := range evals {
+			switch eval.Status {
+			case structs.EvalStatusBlocked:
+				evalStatus = "blocked"
+				break
+			case structs.EvalStatusPending:
+				break
+			case structs.EvalStatusComplete:
+				break
+			case structs.EvalStatusFailed:
+				break
+			case structs.EvalStatusCancelled:
+				break
+			default:
+				evalStatus = "unknown"
+			}
+		}
```
Could this be squashed down to a `hasSquashedEvals` bool? That'd make it more clear what the purpose of this block is and let the check below just be on that bool.

paladin-devops replied:
@tgross will do! I originally wrote this with a bool, but I changed to using a string to account for the unknown case, so we could report an unknown status if there were a new Nomad eval status introduced but not handled here.

tgross reviewed (commented):
This is looking pretty good from a Nomad perspective @paladin-devops.

briancain commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,577 @@
+					select {
+					case <-ticker.C:
+					case <-ctx.Done(): // cancelled
+						return status.Errorf(codes.Aborted, "Healthy deadline reached: %s", ctx.Err())
```
```suggestion
						return status.Errorf(codes.Aborted, "Context cancelled from timeout checking health of task group %q: %s", group, ctx.Err())
```

I think this is more accurate here?

briancain commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,577 @@
+		return err
+	}
+
+	st.Update(fmt.Sprintf("Monitoring evaluation %s", string(u.EvalID)))
```
StepGroup I believe supports string formatting through its parameters, so you shouldn't need `fmt` here, should be able to do `"Monitoring evaluation %q", string(u.EvalID)` without the `fmt.Sprintf`!

briancain replied:
I might of missed this else where, so if there are other places, it would be good to get those updated too!

paladin-devops replied:
Thanks @briancain! Since `st` here is of type `terminal.Status`, did you mean to convert this (and other usages of Status.Update) to use `StepGroup`?

Also just for clarification, I used `terminal.Status` here instead of `StepGroup` as I saw other comments mentioning to not use `StepGroup` until #1536 is resolved.

briancain replied:
Yeah, so this is a bit confusing. This `st.Update` is a StepGroup, and `Status` is apart of that implementation I believe. The issue linked basically says we should avoid setting `Status` after a `StepGroup` output due to that bug, but I don't think we're doing that here! Just some extra info for you right now as to what that issue is about 😄 

But what I was suggesting is this should be possible, no need to use `fmt.Sprintf`:

```golang
st.Update("Monitoring evaluation %s", string(u.EvalID))
```

paladin-devops replied:
@briancain I actually gave that a try, but it wouldn't compile 🤔 resulting in this warning:
```
too many arguments in call to st.Update
	have (string, string)
	want (string)
```

Dug through the core code, and _think_ the root cause is that `runnerUIStatus` (which is what I'm using in `resourceJobCreate`) does not do the formatting for us already...

```go
func (u *runnerUIStatus) Step(status string, msg string) {
	u.b.mu.Lock()
	defer u.b.mu.Unlock()

	if u.b.evc == nil {
		return
	}
```

https://github.com/hashicorp/waypoint/blob/main/internal/runner/ui.go#L255-L262

...but `runnerUISGStep` does!

```go
func (u *runnerUISGStep) Update(str string, args ...interface{}) {
	msg := fmt.Sprintf(str, args...)

	u.sg.ui.mu.Lock()
	defer u.sg.ui.mu.Unlock()
```

https://github.com/hashicorp/waypoint/blob/main/internal/runner/ui.go#L416-L421

I saw examples in other plugins which use basic concatenation with `status.Update` instead of string formatting, could do that here as well! Let me know what you think. 😄 

```go
	st.Update("Deleting Lambda function version " + deployment.Version)
```

briancain replied:
Oops - ok that makes sense. I guess it was originally a `Status` and not a `StepGroup`! I think your example below with the string concatenation makes sense 👍🏻 Sorry for the mixup!

briancain commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,577 @@
+	state *Resource_Job,
+	sg terminal.StepGroup,
+) error {
+	// Do nothing because the platform will destroy the job for us
```
Maybe a `log.Trace` here saying no action happened would be good for debugging.

briancain commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,577 @@
+		jobResource.Health = sdk.StatusReport_MISSING
+		jobResource.HealthMessage = sdk.StatusReport_MISSING.String()
+	} else if job == nil {
+		return status.Errorf(codes.FailedPrecondition, "Nomad job response cannot be empty")
```
Instead of failing the entire status job, we could make this case an UNKNOWN health with this as the health message 🤔

briancain commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,577 @@
+// ReleaserConfig is the configuration structure for the Releaser.
+type ReleaserConfig struct {
+	Groups         []string `hcl:"groups,optional"`
+	FailDeployment bool     `hcl:"fail_deployment,optional"`
```
Variable comments here would be excellent! Could be a copy of the doc comments defined below.

briancain commented on builtin/nomad/jobspec/platform.go:
```diff
@@ -46,86 +54,314 @@ func (p *Platform) GenerationFunc() interface{} {
 
 	// Register job
-	s.Update("Registering job %q...", *job.Name)
+	st.Update(fmt.Sprintf("Registering job %q...", *job.Name))
```
Same feedback here, no need for `fmt`, the StepGroup func should support string formatting!

briancain reviewed (changes requested):
Hey there @paladin-devops !! Thanks for taking on this work and doing such a thorough job of it. I've done a first pass of the code to give you some additional review on the Waypoint side. It's a lot of comments, but this is a big PR! Overall I think this is in the right direction.

I haven't tested it yet, mostly just code review. I will do that kind of review next once all the Nomad/Waypoint feedback has been addressed 😄 Thanks again, well done!

paladin-devops commented:
Thank you for all the help with the Nomad details @tgross, and for taking time to review this while squashing the CSI bugs over there! 😄

& thanks for the 1st pass through @briancain! 😄 I had only one question about your feedback, but I think I got the rest of it covered!

briancain commented:
Looks great @paladin-devops ! You might need to regenerate the website mdx, otherwise you can ignore the Vercel failure right now! I will re-request review for myself so I can give this a proper test 🙏🏻

briancain commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,598 @@
+			groupsToPromote = append(groupsToPromote, taskGroupName)
+		}
+	}
+	if !canaryDeployment {
```
Is this technically a release failure? If there's no canaries defined in the jobspec file, should we abort the release? Right now we say the release was successful in this case.

```
... deploy output ....
✓ Job "web" is reporting ready!

» Releasing example-nodejs...
⚠️ Nomad canary allocations not detected in job task groups.
✓ Release successfully rolled out!

✓ Finished building report for Nomad platform
✓ Finished building report for Nomad job resource
```

briancain replied:
I noticed this happens on the very _first_ run. No canary allocations are detected in the task group, and the second run, they exist 🤔

paladin-devops replied:
Yep, canaries are introduced in the `update` stanza, and since the very first deployment of a job isn't considered an "update", all allocations are deployed right away. I think it'd make sense to add something to check if it's `jobVersion` 0, and then not report an error here!

briancain replied:
That sounds good to me!

briancain commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,598 @@
+			connect {
+				sidecar_service {}
+			}
+		}
```
This job block needs some formatting help. I'm guessing it's just a mix of tabs and whitespace! `network {` on line 490 looks a bit off too.

paladin-devops replied:
You're absolutely right, it's possibly something with the linter in one of the linter plugins I'm using in my IDE (Atom) confusing Go with markdown 😅 should be fixed up now, with the help of good old vim.

briancain commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,598 @@
+		groups = [
+			"app"
+		]
+		fail_deployment = false
```
I'm thinking since this is optional, we don't include it in the docs. If it's unset, it should default to false.

briancain commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,598 @@
+	var currentTaskGroupState *api.DeploymentState
+	for _, group := range groupsToPromote {
+		if group != "" {
+			st.Update(fmt.Sprintf("Checking task group: %s", group))
```
I'd love it if we could also print some health information here from Nomad allocation while we wait. Similar to what we do with the Kubernetes plugin:

```
» Deploying tetris...
✓ Kubernetes client connected to https://127.0.0.1:36367 with namespace default
✓ Creating deployment...
✓ Expected "http" port "3000" for app "tetris-v1"
⠏ Waiting on deployment to become available: requested=1 running=0 ready=0
```

Otherwise for unhealthy deployments, it is unclear what is being checked or why it hasn't progressed. I pasted some extra output from the Nomad CLI that I was looking at while waiting for the task group to come up. I think maybe we could show some of this info from `currentTaskGroupState` if possible.

```
Latest Deployment
ID          = 89389c34
Status      = running
Description = Deployment is running but requires manual promotion

Deployed
Task Group  Auto Revert  Promoted  Desired  Canaries  Placed  Healthy  Unhealthy  Progress Deadline
app         true         false     1        1         3       0        3          2022-02-24T14:18:15-08:00

Allocations
ID        Node ID   Task Group  Version  Desired  Status   Created     Modified
a4a534b0  51dfa101  app         1        run      failed   1m23s ago   43s ago
5da6f961  51dfa101  app         1        stop     failed   3m2s ago    1m23s ago
bfb30a7e  51dfa101  app         1        stop     failed   4m12s ago   3m2s ago
55861890  51dfa101  app         0        run      running  13m21s ago  13m11s ago

Time                       Type             Description
2022-02-24T14:10:03-08:00  Alloc Unhealthy  Unhealthy because of failed task
2022-02-24T14:10:03-08:00  Not Restarting   Exceeded allowed attempts 2 in interval 30m0s and mode is "fail"
2022-02-24T14:10:03-08:00  Terminated       Exit Code: 1, Exit Message: "Docker container exited with non-zero exit code: 1"
2022-02-24T14:10:02-08:00  Started          Task started by client
2022-02-24T14:09:45-08:00  Restarting       Task restarting in 16.755013793s
2022-02-24T14:09:45-08:00  Terminated       Exit Code: 1, Exit Message: "Docker container exited with non-zero exit code: 1"
2022-02-24T14:09:44-08:00  Started          Task started by client
2022-02-24T14:09:26-08:00  Restarting       Task restarting in 18.214921411s
2022-02-24T14:09:26-08:00  Terminated       Exit Code: 1, Exit Message: "Docker container exited with non-zero exit code: 1"
2022-02-24T14:09:24-08:00  Started          Task started by client
```

I haven't figured out why my Nomad deployment was unhealthy and the container kept exiting, it might be an issue on my end! But some UX I noticed could be improved :)

briancain replied:
If you have an idea of why it might be exiting let me know. I don't see any failures in docker logs or Nomad 🤔 I can provide Nomad jobspec and Waypoint HCL files of course if needed.

paladin-devops replied:
@briancain the jobsepc and HCL would definitely help! I can take a look at that for you, & will look into how to improve the UX here.

paladin-devops replied:
Since the pre-requisite for promoting a Nomad canary deployment is that all canary allocations are healthy, I added a line to display the # of healthy allocs vs. the number of deployed canaries. Once they're equal, that's when the promotion can happen!

```go
st.Update("Waiting on allocations to become healthy: healthy allocs=" + strconv.Itoa(currentTaskGroupState.HealthyAllocs) + " placed canaries=" + strconv.Itoa(len(currentTaskGroupState.PlacedCanaries)))
```

And the output looks like (couldn't copy it live from terminal so took screenshot):
![image](https://user-images.githubusercontent.com/83741749/155660167-0ed4393a-07a2-4300-8289-b17a55ccb8f2.png)

briancain replied:
Looks amazing! Love it - thanks for adding that in 👍🏻

briancain reviewed (commented):
OK! I was able to give this a test run. I haven't got a successful Canary promotion yet, but I got through enough of it to give a better review of the pull request.

And major props adding status to jobspec, this is quite a large pull request with a lot of major features, so kudos! ✨

paladin-devops commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,598 @@
+
+	doc.SetField(
+		"fail_deployment",
+		"If true, fails the canary deployment.",
```
> Is it that if it's true, instead of promoting the deployment for release, it fails the deployment by marking it as bad in Nomad? 

^ correct! But agreed that it could be articulated better! I intend to update it to "If true, marks the deployment as failed." - stolen from the Nomad CLI docs! 😄 

https://www.nomadproject.io/docs/commands/deployment/fail#command-deployment-fail

briancain commented on builtin/nomad/jobspec/releaser.go:
```diff
@@ -0,0 +1,605 @@
+
+	rm := r.resourceManager(log, dcr)
+	if err := rm.CreateAll(
+		ctx, log, st, &result, target,
```
This is a pretty subtle error, but we need to include a `StepGroup` parameter here for `CreateAll`. The reason being, for when a Release encounters a failure, Waypoint automatically "rolls back" to a previous version. We do this through the plugin SDK, and end up calling `DestroyAll` directly rather than the top level `Destroy` here. `DestroyAll` expects a stepgroup, and without it, we get a bad error:

```
! 2 errors occurred:
        * rpc error: code = Aborted desc = Context cancelled from timeout checking
  health of task group "app": context deadline exceeded
        * Error during rollback: 1 error occurred:
        * argument cannot be satisfied: type: terminal.StepGroup. This is a bug in the
  go-argmapper library since this shouldn't happen at this point.
```

This is basically saying when our internal SDK attempted to call `DestroyAll` during a Release failure, the arguments it were given could not satisfy any defined Destroy functions that expected a StepGroup.

For the fix, it is pretty simple:

```diff

diff --git a/builtin/nomad/jobspec/releaser.go b/builtin/nomad/jobspec/releaser.go
index 237218fe8..7cb514b10 100644
--- a/builtin/nomad/jobspec/releaser.go
+++ b/builtin/nomad/jobspec/releaser.go
@@ -319,11 +319,12 @@ func (r *Releaser) Release(
        // TODO: Replace ui.Status with StepGroups once this bug
        // has been fixed: https://github.com/hashicorp/waypoint/issues/1536
        st := ui.Status()
+       sg := ui.StepGroup()
        defer st.Close()
+       defer st.Wait()

        rm := r.resourceManager(log, dcr)
        if err := rm.CreateAll(
-               ctx, log, st, &result, target,
+               ctx, log, st, sg, &result, target,
        ); err != nil {
                return nil, err
        }

```

paladin-devops replied:
Thanks Brian! Just pushed the update, good catch!

briancain reviewed (commented):
Yoo, thanks for all the updates @paladin-devops - I still can't get my Canary to successfully work, but loving all the updates. I found a bug in the failure scenario for releases, otherwise this looks great to me.

I've asked some others on the team to help test this PR too and see if they can get it to work. If you have any other advice for getting the Canary successful I'm all ears 😄

briancain commented:
I'm pretty good with the code changes here 👍🏻 I'll give the team a few days to see if others can get it to work on their systems.

I assume a Canary release works on yours @paladin-devops ? Is there something specific from the jobspec that I might need? I really only used the `update` stanza:

```hcl
// Nomad jobspec file app.nomad.tpl
job "web" {
  datacenters = ["dc1"]
  group "app" {
    update {
      max_parallel = 1
      canary       = 1
      auto_revert  = true
      auto_promote = false
      health_check = "task_states"
    }

    task "app" {
      driver = "docker"
      config {
        image = "${artifact.image}:${artifact.tag}"

        // For local Nomad, you prob don't need this on a real deploy
        network_mode = "host"
      }

      env {
        %{ for k,v in entrypoint.env ~}
        ${k} = "${v}"
        %{ endfor ~}

        // For URL service
        PORT = "3000"
      }
    }
  }
}
```

```hcl
project = "example-nodejs"

config {
  env = {
    WP_TEST = "test-string"
  }
}

app "example-nodejs" {
  build {
    use "pack" {}
    registry {
      use "docker" {
        image = "nodejs-example"
        tag   = "1"
        local = true
      }
    }
  }

  deploy {
    use "nomad-jobspec" {
      // Templated to perhaps bring in the artifact from a previous
      // build/registry, entrypoint env vars, etc.
      jobspec = templatefile("${path.app}/app.nomad.tpl")
    }
  }

  release {
    use "nomad-jobspec-canary" {
      groups = [
        "app"
      ]
    }
  }
}
```

paladin-devops commented:
> I assume a Canary release works on yours @paladin-devops ? Is there something specific from the jobspec that I might need? 

@briancain I used your `waypoint.hcl` file and Nomad jobspec template file on my system (Linux amd64, Debian based), along with the code from the `waypoint-examples` repo (I assumed that's what your `pack` build was using), and I was also unable to deploy it! The Docker containers were unhealthy. Here's what I ran into:

From Nomad alloc logs:
```
events.js:291
      throw er; // Unhandled 'error' event
      ^

Error: listen EADDRINUSE: address already in use :::3000
    at Server.setupListenHandle [as _listen2] (net.js:1316:16)
    at listenInCluster (net.js:1364:12)
    at Server.listen (net.js:1450:7)
    at Function.listen (/workspace/node_modules/express/lib/application.js:618:24)
    at Object.<anonymous> (/workspace/index.js:10:4)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)
Emitted 'error' event on Server instance at:
    at emitErrorNT (net.js:1343:8)
    at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  code: 'EADDRINUSE',
  errno: 'EADDRINUSE',
  syscall: 'listen',
  address: '::',
  port: 3000
}
```

So what I'm thinking here is that _possibly_ "network_mode" host is causing this port conflict. I removed that from my template, and it started up! I'd recommend removing that and then trying it out.

Happy result (from Nomad alloc logs):
```
> node-js-getting-started@0.3.0 start /workspace
> node index.js

Listening on 3000
```

> edit: Also, sorry for all of the extra debug work here to get your PR over the finish line. If you do not have time for this at the moment I understand.

No problem! I know all the enhancements will help my fellow Nomads 😄

briancain reviewed (approved):
Thanks again for this! Looks great, appreciate all of the time spent on feedback and review from us.

tgross reviewed (approved):
LGTM 👍

briancain commented:
Let's merge this thing! 🎉
